### ISSUE TITLE
Test failures seen with ASAN enabled on Ubuntu 20.04 running 5.12, 5.13 upstream kernel

### ISSUE THREAD

**Issue by jinengandhi-intel** [gramineproject/gramine#208]
## Description of the problem

Seeing multiple issues when enabling ASAN on local CI Ubuntu 20.04 machines running 5.12 and 5.13 kernel.

**Issue 1: Gramine SGX on Ubuntu 20.04 + 5.12 kernel**

Below error is seen for multiple PAL, PF, LibOS tests. Logfiles for a few of the tests are attached here. When run via pytest the test is hung and eventually times out.

[LIBOS_mprotect_filefork_SGX_ASAN.txt](https://github.com/gramineproject/gramine/files/7503628/LIBOS_mprotect_filefork_SGX_ASAN.txt)
[LIBOS_forandexec_SGX_ASAN.txt](https://github.com/gramineproject/gramine/files/7503629/LIBOS_forandexec_SGX_ASAN.txt)
[PAL_Process_ASAN_SGX.txt](https://github.com/gramineproject/gramine/files/7503630/PAL_Process_ASAN_SGX.txt)
[asan_gsgx_fail_test-list.txt](https://github.com/gramineproject/gramine/files/7503718/asan_gsgx_fail_test-list.txt)
```
error: Unexpected memory fault occurred inside PAL (0xba3fc1f)
debug: (untrusted PAL sent PAL event 0x2)
debug: DkProcessExit: Returning exit code 1
```

**Issue 2: test_asan_buffer_overflow failing intermittently with Gramine-direct on Ubuntu 20.04 + 5.12 kernel**
Except for this 1 test, all LibOS, PAL, FS tests run without any issues. Log file for the failing  run is attached.
[LibOS_ASAN_unit_test.txt](https://github.com/gramineproject/gramine/files/7503655/LibOS_ASAN_unit_test.txt)

**Issue 3: Gramine-direct tests hang on Ubuntu 20.04 + 5.13 upstream kernel**

Prompt is never returned and even Ctrl + C doesn't work, the process has to be killed from a new session. On the same machine Gramine-SGX exhibits the same behavior as 5.12 kernel i.e. passes for some tests and fails for some.
 

## Steps to reproduce

Build Gramine with ASAN enabled and run the tests. Let me or @aniket-intelx   know if you need access to the specific machine where we see the issues.

**Comment by pwmarcz** [comment]
I see there's no information about crash location in any of these logs. Is Gramine compiled in debug mode, or release mode? What is the full `meson` invocation?

**Comment by jinengandhi-intel** [comment]
Meson command: `meson build --prefix=/home/intel/gramine_install/usr --buildtype=release -Dskeleton=enabled -Ddirect=enabled -Dtests=enabled -Dubsan=enabled -Dasan=enabled -Dsgx=enabled`

**Comment by pwmarcz** [comment]
I see. Then at least `Issue 2` looks like a simple typo in the Python code (`has_debug` -> `has_debug()`). I'll submit a fix.

Do the same issues appear when compiling in debug mode (`--buildtype=debug` or `--buildtype=debugoptimized`)? If so, that could give us more information about the crashes.

**Comment by aniket-intelx** [comment]
Checked out the latest master (which already has your fix) and as you mentioned issue 2 is no longer seen when run in release mode.
For issue 1, we build in debug mode and the memory fault is no longer seen. Is this logical?
For issue 3, which is tests hanging when run with gramine-direct, the issue is still seen even with debug mode. We tried with log_level as `trace`, but no message is seen on console and it simply hangs.
Please let us know if you need access to this machine where this issue is seen.

**Comment by jinengandhi-intel** [comment]
@aniket-intelx can you apply the following diff and re-run gramine-direct on 5.13 kernel.

**Comment by aniket-intelx** [comment]
> @aniket-intelx can you apply the following diff and re-run gramine-direct on 5.13 kernel.

`gramine-direct`, tested with LibOS and PAL testcases is no longer hanging after applying the above mentioned fix.

**Comment by dimakuv** [comment]
@aniket-intelx @jinengandhi-intel Is this issue still relevant? Any of the sub-issues 1, 2, 3 not yet fixed?

**Comment by jinengandhi-intel** [comment]
Issue 2 and 3 are fixed but Issue 1 is still relevant, which we see memory fault error when run in release mode. Pawel, asked us to run in debug mode and that seemed to have fixed the issue but doesn't sound like a solution.

**Comment by pwmarcz** [comment]
You mentioned that with debug mode, the issue no longer appears. How about `debugoptimized`? If it appears there, we might get more information (especially since we improved error reporting recently).

**Comment by jinengandhi-intel** [comment]
@aniket-intelx like Pawel asked could you try with `--buildtype=debugoptimized` and share the results with Pawel?

**Comment by aniket-intelx** [comment]
The steps followed while building are given below.
```
export CC=clang CXX=clang++ AS=clang
export ASAN=1
export UBSAN=1
meson setup build  --buildtype=debugoptimized -Ddirect=enabled -Dtests=enabled -Dubsan=enabled -Dasan=enabled -Dsgx=enabled -Dmusl=disabled
```
Please find the attached logs for different suites.
[FS_Gdirect_and_SGX.txt](https://github.com/gramineproject/gramine/files/8021436/FS_Gdirect_and_SGX.txt)
[LIBOS_Gdirect_and_SGX.txt](https://github.com/gramineproject/gramine/files/8021437/LIBOS_Gdirect_and_SGX.txt)
[PAL_Gdirect_and_SGX.txt](https://github.com/gramineproject/gramine/files/8021438/PAL_Gdirect_and_SGX.txt)

**Comment by pwmarcz** [comment]
For the record:

```
E   subprocess.CalledProcessError: Command '['gramine-sgx-pf-crypt', 'decrypt', '-w', 'tmp/wrap-key', '-i', 'tmp/pf_output/0', '-o', 'tmp/pf_output/0.dec']' died with <Signals.SIGSEGV: 11>.

error: Unexpected memory fault occurred inside PAL (_DkStreamReportRespond at enclave_framework.c, libpal.so+0x7eaa3)
```

**Comment by pwmarcz** [comment]
Thanks @jinengandhi-intel @aniket-intelx! I think I found out the cause of both crashes, see the attached items (#389, #391).

**Comment by pwmarcz** [comment]
I think that after merging #389 and #398 all problems described in this issue have been fixed, please verify.

**Comment by aniket-intelx** [comment]
> I think that after merging #389 and #398 all problems described in this issue have been fixed, please verify.

We are observing failure for `test_022_asan_stack` and `test_023_asan_stack` in LibOS suite for both `gramine-direct` and `gramine-sgx`.

The steps we followed while building are given below.
```
export CC=clang CXX=clang++ AS=clang
export ASAN=1
export UBSAN=1
meson setup build  --buildtype=release -Ddirect=enabled -Dtests=enabled -Dubsan=enabled -Dasan=enabled -Dsgx=enabled -Dmusl=disabled
```
Kindly let us know if we are missing something.
The log for LibOS suite is attached below.
[LIBOS_ASAN.txt](https://github.com/gramineproject/gramine/files/8066254/LIBOS_ASAN.txt)

**Comment by pwmarcz** [comment]
Thanks for testing. You're right: you reported these before and I forgot. I submitted a fix in #410.

**Comment by mkow** [comment]
Seems to be long fixed, so I'm closing it.
